### DISCUSSION TITLE
Multiple geoPoints per document

### DISCUSSION THREAD

**Discussion by gmourier** [meilisearch/product#508] [Feedback & Feature Proposal]
Allowing to store several geo coords per document could allow finding a document in case its geo coords are in several shapes or one or several.

Please, tell us why this could be interesting for your use case!

Thank you!

**Comment by mmachatschek** [comment]
We have an index for companies that have multiple subsidiaries where we have the addresses and geopoints that we would like to search through.

ATM we need to add a workaround to search for all the geopoints by either 
- indexing the companies multiple times with different primary keys and different subsidiaries (in conjunction with https://docs.meilisearch.com/reference/api/distinct_attribute.html#get-distinct-attribute)
- or by separating the company subsidiary locations into a separate index and search the geopoints there, get the company ids from the result and filter the companies by the ids in the main `companies` index

**Reply by gmourier** [reply]
Thanks @mmachatschek

**Reply by jhf** [reply]
When using typing completion, we would like to provide hits for relevant facets of documents. To do this we need a facet index, where each facet has the geopoints of the related documents. This allows our completion to be geographically relevant.

**Reply by gmourier** [reply]
Hey @jhf, sorry to be late 👋

I'm not sure I understand well. Could you share an example of the data and a mock-up/screen of the interface to see how you would interact with it?

Thanks!

**Reply by furmannodesol** [reply]
Any solution for your multiple subsidiaries? @mmachatschek

**Reply by mmachatschek** [reply]
I currently use my first suggested solution from above. (Indexing the model multiple times with a custom key and using a distinct attribute) current downside we have is that the estimatedHits count is not really accurate

**Reply by furmannodesol** [reply]
Thank you @mmachatschek for the reply.
How do you fetch the results with a custom key which actually doesn't exist in the database. Would be great if you could elaborate little bit more.

**Reply by jhf** [reply]
> Hey @jhf, sorry to be late 👋
> 
> I'm not sure I understand well. Could you share an example of the data and a mock-up/screen of the interface to see how you would interact with it?
> 
> Thanks!

Hi!
So, we have a search function, where the user may search for something, say the name of a place, the name of a shop, the name of a brand, the name of a category for a shop. However these results are always limited by geoposition, so the search may be done nearby London, or som place with a geographical position.
As a simplified example, say there are shops in the database:
* Shop A - Sporting goods - Near London
* Shop B - Food - Near London
* Shop D - Food - Far away from London.
* Shop C - Shoes - Far away from London.

Now the user enters the free text search, and types "S".
We would like to make a suggestion of completions, and that list could look like:
* "Sporting Goods"
* "Shop A"
* "Shop B"

And in that list, there should not be any "Shoes" because there are no relevant search results for that nearby London.

However if the user starts typing "F" then the suggestions may be
* "Food"

So how could we organise the data for searching?

I've assumed that we can not use some kind of nested data structure
* One index for the shops, each with one geoposition.
* Add the category "Shoes"/"Food"/"Sporting Goods" to the shop.
Could we search for completion using one index?
Could one make another index for category inside shop?

Rather we create different indices, one for shop and another for category.
* Shop index with name and geoposition.
* Category index with name and list of geopositions.
And then search over two indices and order according to say distance, and within some distance interval by relevance.

Does that make it clearer @gmourier ?
Oh, and thank you for taking the time to ask!

Now the

**Comment by alexphiev** [comment]
Great proposal! Ideally, something like Algolia's solution: 
```
"_geoloc": [
  { "lat": 47.279430, "lng": 5.106450 },
  { "lat": 47.293228, "lng": 5.004570 },
  { "lat": 47.316669, "lng": 5.016670 }
]
```
Reference: https://www.algolia.com/doc/guides/managing-results/refine-results/geolocation/

Thanks a lot!

**Comment by mmachatschek** [comment]
Hey @gmourier, sorry for pinging you but are there any updates or further thoughts on this topic?

**Reply by gmourier** [reply]
Hello @mmachatschek 👋

Unfortunately, no strong advancements to share, we have other focuses to concentrate on at the moment. I had a quick discussion with @irevoire and this may be something that we can consider a quick win, maybe we could open up a good first issue to see if someone solves it until it fits into our pipeline.

**Reply by mmachatschek** [reply]
thanks for sharing the current status 🙌  @gmourier

**Reply by mmachatschek** [reply]
Hi @gmourier I would be happy to see the issue on the meilisearch/meilisearch repo. This would probably be something that I would try to explore and implement myself or to do a POC.

**Reply by irevoire** [reply]
Hey @mmachatschek, today I spent some time trying to describe how we could implement it, but there is a flaw in our filter design.
Now, since we have the `_geoRadius` operator I don't think we can implement it without rewriting a considerable part of the engine either on the filtering part (that would solve a second issue we have with the filter around nested objects) or a big rewrite of the geodatabase (that would be the occasion to handle the geo shape).
Whatever happens, I was very wrong when I said it was a quick win. Sorry, it's actually very hard to implement, and I can't even estimate which solution we should choose and how long it would take sorry 😢 
Next monday I'll talk about it with kero and come back to you if he has a better idea.

**Reply by mmachatschek** [reply]
@irevoire very appreciate the effort here 👍

**Reply by irevoire** [reply]
Ok so in conclusion it's as I said.
One “not too long fix” would be to introduce a new special database just to handle the geo bounding box, but I don't think we're going to do that  gr
We're probably going to wait until we rewrite the whole geo pipeline 😖

**Comment by ramanjitsevo** [comment]
**Multiple** **geoPoints** for a single document is it still an issue or it has been fixed?
My data is something like:
   "locations" => [
      0 => array:[
        "id" => 8894
        "lat" => 33.97
        "lng" => -84.36        
      ],
      1 => array:[
        "id" => 579723
        "lat" => 33.7382
        "lng" => -84.4079        
      ],
      2 => array:[
        "id" => 579723
        "lat" => 33.7382
        "lng" => -84.4079        
      ]
    ]
A single documents contains multiple locations.
I need to find documents within 20 miles range from $lat and $lng.

**Reply by gmourier** [reply]
Hey @ramanjitsevo 👋

While we are committed to other priorities right now, a workaround may be worth evaluating on your side in the meantime even if not ideal.

It has previously been described by @mmachatschek [here](https://github.com/meilisearch/product/discussions/508#discussioncomment-3348157)

You could split your different locations into multiple documents and denormalize the root data, such as brand/company info within its various documents.

```jsonc
[
     {
          "id": "7adff1a6-20cf-11ee-be56-0242ac120002", //the id of a location; could be auto-generated or hashed from the lat, lng if there is no location identifier at hand
          "_geo": {
              "lat": 33.97,
              "lng": -84.36
          },
          ...,
          "brand": "Burger King",
          "brand_id": 1
    },
    {
          "id": "eb197582-20cf-11ee-be56-0242ac120002",
          "_geo": {
              "lat": 33.7382,
              "lng": -84.4079
          },
          ...,
          "brand": "Burger King",
          "brand_id": 1
    },
    {
          "id": "f73b52b8-20cf-11ee-be56-0242ac120002",
          "_geo": {
              "lat": 32.7382
              "lng": -83.4079
          },
          ...,
          "brand": "Taco Bell"
          "brand_id": 2
    }
]
```

Have you considered it?

Thanks

**Reply by martin-landhi** [reply]
Sound good temporally solution. i will test it and come back in a few days with conclutions, thanks @gmourier

**Reply by mmachatschek** [reply]
@martin-landhi, we are using this feature in production for ~ 9months. It is a solution but (hopefully only temporarily, until multi geo point support has landed) its slow (if you use paginated results) and needs some extra implementation to keep all documents for each variant in sync. Especially empty queries (landingpages) are very slow if you don't cache search results

empty queries -> super slow (if you have many variants)
using a search query + filters -> gets faster if less documents are in the search result (meilisearch doesn't have to do too much computation for the subset)

**Reply by martin-landhi** [reply]
Perfect @mmachatschek i will test it with more or less 20k documents and see the performance, but caching the answer can be good solution if it is slow, because most of them will be empty queries

**Reply by martin-landhi** [reply]
i could apply your solution @mmachatschek, thanks. About the performance right now in not an issue but when the data growth maybe i have to apply cache for this.

**Comment by martin-landhi** [comment]
So, we should not expect any progress here soon?

I have also many location per document to be searcheables

**Reply by ramanjitsevo** [reply]
Waste of time. I had to move to another platform.

**Reply by macraig** [reply]
Hi @martin-landhi @ramanjitsevo 👋 Unfortunately, we haven't made progress on this front. We truly understand how important this feature is to you and recognize its potential value for our users. While we'd love to address all user requests promptly, there are times when we face challenges that require us to reprioritize our efforts. Currently, our team is focused on addressing critical issues and working on features that have a broader impact across our user base.

Please know that we take your feedback and requests seriously, and we deeply appreciate your patience and understanding in this matter. We want to assure you that your requests have been duly noted, and we have not abandoned the idea. While we don't have a concrete timeline to share for the implementation at this moment, we assure you that we will revisit it as soon as our current priorities allow. Your feedback and input are valuable to us, and we are committed to continually improving our product to better meet your needs. We appreciate your ongoing support and look forward to serving you with a more comprehensive product experience in the future.

**Comment by macraig** [comment]
Adding feedback from @martin-landhi in https://github.com/meilisearch/product/discussions/681 

> Is meilisearch Geolocation feature allow have several locations per document?
For example:
I have a companies that have projects in several locations. I want to search for companies that are working on some Location.
Meilisearch allow to create the document for those comapanies with all the locations where they are working and then search?

**Reply by furmannodesol** [reply]
I have implemented it with creating a custom primary ID (which is actually not a database column), I was able to do that and achieving my goals. However it was working fine for up to version v0.30. 
However unable to create documents after upgraded to v1.2. Getting message as number of X have been imported and the number of actually created document shows ZERO in the UI.
Any help please

**Reply by Kerollmops** [reply]
Hey @furmannodesol 👋

Could you explain what you were doing and that no more works in the v1.0+ versions of Meilisearch by showing an example document update you are sending, please?

I'm not sure to understand. Were you creating a document for each geo-point with the original document id (set as distinct) and doing another search to retrieve the original documents?

Have a nice day 🍡

**Reply by furmannodesol** [reply]
Hello @Kerollmops custom
I have created indexes with custom primary key like
php artisan scout:index products --key custom_primary_key
WHERE
custom_primary_key is not a database table field.
and it works fine until v0.30 as I can add multiple documents to the index having same ID but with different custom_primary_key
HOWEVER
when I upgraded meilisearch to v1.2, it stops adding documents. I can see X records imported in the command line message but actually there are 0 product added. And only works if I replace custom_primary_key with any other unique field from database like ID or SLUG.

**Reply by Kerollmops** [reply]
Hey @furmannodesol,

Your issue seems more related to the primary key inference than the multi-geo point feature. We revamped the inference of the primary keys in v1.0.0. You can search for _Revamp primary key inference_ [on the release page](https://github.com/meilisearch/meilisearch/releases/tag/v1.0.0) to understand the difference, but basically, I advise you to define the primary key when creating your index.

If you have questions on how to do it with Laravel Scout, please, ask your question [on our Discord to get help](https://discord.meilisearch.com/).

**Reply by mmachatschek** [reply]
@furmannodesol please also check out if your laravel/scout version is compatible with your meilisearch version there was an major version bumpt on scout side that added support for meilisearch v1. if you didn't upgrade the laravel/scout package you may  experience issues because of incompatibility

**Reply by furmannodesol** [reply]
Hello @mmachatschek 
Thanks for responding here!
In fact I have 
Scout: v10.2.3
Meilisearch PHP package: v1.2.1
Meilisearch: v1.2.1

**Comment by martin-landhi** [comment]
I could made this appling several documents per item with multiple address and applying district id as described above.

Thanks

**Comment by garygreen** [comment]
Also sharing our interest in this feature. Our use case: we have members for hire on our site who set multiple areas they operate in, and we need a way for those to be discovered by others around a given point.

We tried to move to MeiliSearch from ElasticSearch, but found this key missing feature. The workarounds of having multiple indexes/copies of the data is not workable for us, which is unfortunate. ES, Algolia, RediSearch all support multiple location search, so MeiliSearch is lagging behind in this aspect.

I would love to see this rated higher in priority.

**Comment by mhdismail** [comment]
Any updates or a possible solution?

It's actually crucial to have support for multiple geo locations. We have a nested array of objects, and we want to find all items if at least of the objects (nested) is nearby by a certain threshold.

**Comment by mmachatschek** [comment]
For everyone interested in this topic.
Please upvote this discussion with the (upvote) button (if not already done) on the initial message of [gmourier](https://github.com/gmourier). This will help bumping the priority of this feature

**Comment by electronick86** [comment]
Hello, Is there any plan to develop such a feature in Meilisearch?

At the moment I use the technique of duplicating the documents in the index and using a distinct to only get one document in the result. The downside is that my facets counts are wrong as they don't take the distinct into account.

It would be great to avoid this work of creating several documents.

**Reply by macraig** [reply]
Hello @electronick86 , we do plan to add this feature in the future as we recognize its value, but at the moment we don't have a specific timeline for its release. In the meantime, your input helps us prioritize it, so we really appreciate you bringing this up.

**Comment by soulshake** [comment]
Adding our use case: we have a database of historical postcards, each of which can be associated with an arbitrary number of places (origin, destination, etc). Our users want to search for an address and find all postcards related to that place.

Really hoping to see this feature added!

**Reply by macraig** [reply]
Hi @soulshake, thanks for sharing your use case! As I mentioned in a comment above, we haven't prioritized this feature just yet, but your vote and comment help us prioritize it.

**Reply by irevoire** [reply]
Hey @soulshake, thanks a lot for describing your use-case, I'm answering here to avoid flooding the other post!

> In our case (a database of historical postcards), each postcard can be associated with multiple, completely unrelated places.

Awesome, I'm so sad my city have _no_ postal card registered in your database ahah

> For example, a postcard sent from Paris to San Francisco should appear as a result in searches for Paris, or France, or San Francisco, or California, etc.

I understand the use case, but I don't understand how it works.
Let's say you search for "France", what is happening in reality?
I played a bit on your website, and my guess is that you never really "geo-search" for Paris, but only the bounding box of where the map is, right?
And when we click on Paris on the left, it's just a filter _if_ the label has been correctly inputted?

If that's the case, then you can just convert your `_geoBoundingBox` to a `_geoPolygon` containing each of the corners of your screen I believe 🤔

**Reply by soulshake** [reply]
> Awesome, I'm so sad my city have _no_ postal card registered in your database ahah

Well now I'm curious which city! :)

> Let's say you search for "France", what is happening in reality?
> I played a bit on your website, and my guess is that you never really "geo-search" for Paris, but only the bounding box of where the map is, right?

Yes, exactly. In more detail: When you search for a place and select one of the autocomplete results (from the google places autocomplete API), and the map is centered on the coordinates for that place. Then it does a `POST` to `/multi-search` with a payload like:

```
{
  "queries": [
    {
      "indexUid": "production_PlaceLink",
      "filter": [
        "_geoBoundingBox([53.4357192066942, 21.489257812500004], [38.272688535980976, -16.347656250000004])"
      ]
    }
  ]
}
```

...where `_geoBoundingBox` represents the current view of the map (which can then be panned/zoomed/etc).

> And when we click on Paris on the left, it's just a filter if the label has been correctly inputted?

Yes, exactly. The places in our database are stored hierarchically and I'm using Meilisearch's hierarchical faceted search feature for the filters on the left, along with the `_geo` field to display results on a map. (The `PlaceLink` table is the source of truth for both of these.) There's some duplication there, but I really wanted users to have both options: 1) explore via an interactive map, and 2) drill down via a hierarchy of places (so they can subscribe to notifications for a given city/region/etc).

> If that's the case, then you can just convert your _geoBoundingBox to a _geoPolygon containing each of the corners of your screen I believe 🤔

Oh I see what you mean now! So with your proposed change, instead of indexing PlaceLink, I would only need an index for Postcard, and a serialized Postcard would look something like the following (redacted for legibility)?:

```
{
  "id": 441296,
  "description": "Buena Vista Ate Dinner @ Western Log Cabin Cafe~Went to Continental Divide~1950",
  "recipient": "Mrs. Ray T. Miller",
  "sender": "The Sharps",
  "year": "1950",
  "_geojson": {
    "type": "FeatureCollection",
    "features": [
      {
        "type": "Feature",
        "properties": {
          "id": 6635,
          "types": [ "locality", "political" ],
          "categories": [ "visual" ],
          "name": "Buena Vista",
          "formatted_address": "Buena Vista, CO 81211, USA"
        },
        "geometry": {
          "type": "Point",
          "coordinates": [
            -106.131129,
            38.842218
          ]
        }
      },
      {
        "type": "Feature",
        "properties": {
          "id": 3777,
          "types": [ "locality", "political" ],
          "categories": [ "origin" ],
          "name": "Colorado Springs",
          "formatted_address": "Colorado Springs, CO, USA"
        },
        "geometry": {
          "type": "Point",
          "coordinates": [
            -104.821363,
            38.833882
          ]
        }
      },
      {
        "type": "Feature",
        "properties": {
          "id": 19449,
          "types": [ "political", "locality" ],
          "categories": [ "destination" ],
          "name": "Waveland",
          "formatted_address": "Waveland, IN 47989, USA"
        },
        "geometry": {
          "type": "Point",
          "coordinates": [
            -87.044455,
            39.876988
          ]
        }
      }
    ]
  }
}
```


<details>
<summary>Or the full version</summary>

```
{
  "id": 441296,
  "description": "[ebay listing 297195127237] Buena Vista Ate Dinner @ Western Log Cabin Cafe~Went to Continental Divide~1950",
  "recipient": "Mrs. Ray T. Miller",
  "sender": "The Sharps",
  "source": "ebay_297195127237",
  "created_at": 1745587665.672549,
  "updated_at": "2025-07-27T17:17:46.132",
  "user": [
    "RefriedJeans"
  ],
  "year": "1930",
  "completeness": 5,
  "tx_status_dates": "completed",
  "tx_status_languages": "completed",
  "tx_status_names": "completed",
  "tx_status_places": "completed",
  "tx_status_writing": "unknown",
  "collections": [
    {
      "id": 393,
      "label": "refried.jeans [eBay]"
    }
  ],
  "languages": [
    {
      "id": 17,
      "label": "English"
    }
  ],
  "tags": [],
  "images": [
    {
      "id": 509392,
      "etag": "6127fe6835034743445a7e0415205d8e",
      "phash": "ffff3f01019fffff",
      "created_at": "2025-04-25T13:27:48.547194",
      "text": "",
      "rotation": 0
    },
    {
      "id": 509393,
      "etag": "b8bf141e648f5636606d765990ffaf8c",
      "phash": "40fc301c5f1f5f1e",
      "created_at": "2025-04-25T13:27:53.902743",
      "text": "",
      "rotation": 0
    }
  ],
  "numeric_year": 1930,
  "for_sale": false,
  "_geojson": {
    "type": "FeatureCollection",
    "features": [
      {
        "type": "Feature",
        "properties": {
          "id": 6635,
          "types": [
            "locality",
            "political"
          ],
          "categories": [
            "",
            "visual"
          ],
          "name": "Buena Vista",
          "formatted_address": "Buena Vista, CO 81211, USA",
          "num_ephemera": 4,
          "num_recursive_ephemera": 4,
          "num_followers": 0,
          "ancestors": [
            "United States",
            "Colorado",
            "Chaffee County",
            "Buena Vista"
          ],
          "country": "United States",
          "country_code": "US",
          "country_flag": "/static/flags/us.gif"
        },
        "geometry": {
          "type": "Point",
          "coordinates": [
            -106.131129,
            38.842218
          ]
        }
      },
      {
        "type": "Feature",
        "properties": {
          "id": 3777,
          "types": [
            "locality",
            "political"
          ],
          "categories": [
            "",
            "origin"
          ],
          "name": "Colorado Springs",
          "formatted_address": "Colorado Springs, CO, USA",
          "num_ephemera": 48,
          "num_recursive_ephemera": 59,
          "num_followers": 0,
          "ancestors": [
            "United States",
            "Colorado",
            "El Paso County",
            "Colorado Springs"
          ],
          "country": "United States",
          "country_code": "US",
          "country_flag": "/static/flags/us.gif"
        },
        "geometry": {
          "type": "Point",
          "coordinates": [
            -104.821363,
            38.833882
          ]
        }
      },
      {
        "type": "Feature",
        "properties": {
          "id": 19449,
          "types": [
            "political",
            "locality"
          ],
          "categories": [
            "",
            "destination"
          ],
          "name": "Waveland",
          "formatted_address": "Waveland, IN 47989, USA",
          "num_ephemera": 1,
          "num_recursive_ephemera": 1,
          "num_followers": 0,
          "ancestors": [
            "United States",
            "Indiana",
            "Montgomery County",
            "Waveland"
          ],
          "country": "United States",
          "country_code": "US",
          "country_flag": "/static/flags/us.gif"
        },
        "geometry": {
          "type": "Point",
          "coordinates": [
            -87.044455,
            39.876988
          ]
        }
      }
    ]
  },
  "ebaylisting": {
    "id": 423567,
    "seller": {
      "id": 47,
      "seller": "refried.jeans"
    },
    "price": 9.49,
    "listing_status": "Completed",
    "item_id": 297195127237,
    "sold": false,
    "title": "Buena Vista Ate Dinner @ Western Log Cabin Cafe~Went to Continental Divide~1950",
    "view_item_url": "https://www.ebay.com/itm/297195127237?mkevt=1&mkcid=1&mkrid=711-53200-19255-0&campid=5338957751&customid=REFERENCE_ID&toolid=10050",
    "start_time": 1744166281,
    "end_time": 1744771081,
    "sku": null,
    "healthy": true,
    "is_404": true,
    "superseded_by": null
  },
  "hPlaces": [
    { "lvl0": "United States" },
    { "lvl1": "United States > Colorado" },
    { "lvl2": "United States > Colorado > El Paso County" },
    { "lvl3": "United States > Colorado > El Paso County > Colorado Springs" },
    { "lvl0": "United States" },
    { "lvl1": "United States > Colorado" },
    { "lvl2": "United States > Colorado > Chaffee County" },
    { "lvl3": "United States > Colorado > Chaffee County > Buena Vista" },
    { "lvl0": "United States" },
    { "lvl1": "United States > Indiana" },
    { "lvl2": "United States > Indiana > Montgomery County" },
    { "lvl3": "United States > Indiana > Montgomery County > Waveland" }
  ]
}
```

</details>

And then on the frontend I replace `_geoBoundingBox` with `_geoPolygon` and make `multi-search` point to the other index?

On a related note, it would be :sparkles: amazing :sparkles: if search results could be returned ordered by distance from the center of `_geoPolygon`. The current behavior with `_geoBoundingBox` is confusing if the search results are paginated, because zooming in causes new markers to appear in seemingly random places, if that makes sense.

**Comment by JonathanVorich** [comment]
Nearly one year later still no Update/Progress on this Feature? 
-> the most important usage for me is radius search ...

**Comment by JonathanVorich** [comment]
@gmourier is this feature/rewrite still not in timeline? would be nice to know if it will be in meilisearch in a month, a year or 10 years, so we know if we have to build workaround or switch searchengine 🙂

**Reply by irevoire** [reply]
Hey! We finally made progress on this feature, take a look at my message [here](https://github.com/orgs/meilisearch/discussions/508#discussioncomment-13880344)

**Comment by irevoire** [comment]
Hey everyone! 👋 

I just released a prototype that may fit your use-case: [`prototype-cellulite-1`](https://github.com/meilisearch/meilisearch/pull/5758).
It supports:
- Storing geojson (including multi-points) at indexing time
- Retrieving anything in the database from a polygon specified in a filter (`_geoPolygon([lat,lng],[lat,lng],[lat,lng],...)`)
- ⚠️ It doesn't support sorting yet
- ⚠️ It doesn't work with anything you stored in your current `_geo` field

Read the description for more information, and let us know if this has been useful to you and if we should finish the work.

**Reply by bradjonesca** [reply]
Absolutely useful! Please finish the work.

**Reply by irevoire** [reply]
Hey @bradjonesca, thanks for your interest!
Currently, we're not planning to invest more work on this until we've received more feedback and an example of a use case.
So feel free to share what you're doing with this prototype, and help us find the bugs and the most important feature missing for your use case!
That'll help us prioritize what to focus on next

**Reply by soulshake** [reply]
@irevoire huge thanks for working on this. It would be _extremely_ useful for [our use case described above](https://github.com/orgs/meilisearch/discussions/508#discussioncomment-10990452), which you can see in action [here](http://ephemerasearch.com/search/).

In our case (a database of historical postcards), each postcard can be associated with multiple, completely unrelated places. For example, a postcard sent from Paris to San Francisco should appear as a result in searches for Paris, or France, or San Francisco, or California, etc.

I'm not sure how the polygon search filter would help in our specific case, but multiple geopoints per document would be a game changer. We had to add a lot of extra complexity when migrating from Algolia because Algolia supports this but Meilisearch doesn't. As a workaround, we had to add an extra index for the connections between postcards and locations (i.e. "PlaceLink"), and de-duplicate the results at search time (since a postcard might have the same city as both the origin and the destination). Now, each time a postcard is modified, we have to re-index not only the postcard itself, but every "PlaceLink" document as well -- and vice versa (because we track the number of postcards associated with each place). Keeping all of these documents up to date with every change is very costly in compute and complexity.

Really hopeful to see this feature move forward!

**Reply by JonathanVorich** [reply]
Oh awesome, im looking forward to testing it in my environment and give feedback afterwards!

**Reply by JonathanVorich** [reply]
Looks good, except with no sorting it doesnt really makes sense for us ...

**Comment by remihuigen** [comment]
The lack this feature prevents us from switching to meilisearch from Algolia 😞

Our use case is a little different though from what is already discussed:

Our records represent organisations that are not bound to a single location, but an area (Polygon). This is the service area of the organisation.

Preferably, we would filter the index for X distance to the polygons bounds for any given location (i.e. the users location). If the user is within the Polygon, distance is equal to 0

**Comment by jhf** [comment]
I've opened a PR that adds `_geo_list` support with both **filtering and sorting**: https://github.com/meilisearch/meilisearch/pull/6187

This is a simpler alternative to the GeoJSON approach — you store multiple discrete points as `"_geo_list": [{"lat": ..., "lng": ...}, ...]`, and existing `_geoRadius`/`_geoBoundingBox` filters plus `_geoPoint():asc/desc` sorting just work. Sorting uses the closest (or farthest) point per document, and `_geoDistance` in results reflects the minimum distance across all points.

This addresses the sorting gap mentioned by @JonathanVorich above.

**Comment by jr-cobweb** [comment]
I also need this
